### PR TITLE
Add Biwi Kinect Head Pose dataset.

### PR DESCRIPTION
**Pull request by dnaveenr**
This PR adds the Biwi Kinect Head Pose dataset.

Dataset Request : Add Biwi Kinect Head Pose Database [#3822](https://github.com/huggingface/datasets/issues/3822)

The Biwi Kinect Head Pose Database is acquired with the Microsoft Kinect sensor, a structured IR light device.It contains 15K images of 20 people with 6 females and 14 males where 4 people were recorded twice.

For each frame, there is :
- a depth image, (.bin file)
- a corresponding rgb image (both 640x480 pixels),
- annotation ( present inside a .txt file)
 The ground truth is the 3D location of the head and its rotation.

The dataset structure is as follows : 
```
- 01.obj
- 01
     - frame_00003_depth.bin
     - frame_00003_pose.txt
     - frame_00003_rgb.png
     .
     .
     .
- 02.obj
 - 02
     - frame_00003_depth.bin
     - frame_00003_pose.txt
     - frame_00003_rgb.png
     .
     .
     .
```
Preview of frame_00003_pose.txt : 
```
0.988397 0.0731349 0.133128 
-0.0441539 0.976945 -0.208876 
-0.145334 0.200575 0.968838 

126.665 40.4515 876.198 
```
I have used the following dataset features : 
```
features=datasets.Features(
    {
        "person_id": datasets.Value("string"),
        "frame_number": datasets.Value("string"),
        "depth_image": datasets.Value("string"),
        "rgb_image": datasets.Image(),
        "3D_head_center": datasets.Array2D(shape=(3, 3), dtype="float"),
        "3D_head_rotation": datasets.Value("float"),
    }
```
I am giving the path to the depth_image here.

I need some inputs for the following : 

1. For each person, the dataset has the following additional information : 
```
For each sequence, the corresponding .obj file represents a head template deformed to match the neutral face of that specific person. [*.obj file]
In each folder, two .cal files contain calibration information for the depth and the color camera, e.g., the intrinsic camera matrix of the depth camera and the global rotation and translation to the rgb camera.
```
Wanted to know how we can represent these features ?

2. For _generate_examples , do I parse the directories and fetch the required information ? This would mean reading the .txt file to obtain the "3D_head_center" and "3D_head_rotation" details. We could precompute the features information  and have a metadata file and use the metadata file to yield information in _generate_examples ? Wanted your thoughts for the best approach for this ?

### CODE REVIEW

**Comment by dnaveenr** [comment]
Thanks for the suggestions @mariosasko, yielding one example for each person would make things much easier.
Okay. I'll look at parsing the files and then displaying the information.

**Comment by dnaveenr** [comment]
Added the following : 
- Features, I have included sequence_number and subject_id along with the features you had suggested.
- Tested loading of the dataset along with dummy_data and full_data tests.
- Created the dataset_infos.json file.

To-Do :
- [ ] Update Dataset Cards with more details.
- [ ] "Data Processing" section

Any inputs on what to include in the "Data Processing" section ?

**Comment by dnaveenr** [comment]
@mariosasko  Please could you review this when you get time. Thank you.

**Comment by dnaveenr** [comment]
In the Data Processing section, I've added example code for a compressed binary depth image file. Updated the Readme as well.

**Comment by dnaveenr** [comment]
@mariosasko / @lhoestq , Please could you review this when you get time. Thank you.

**Comment by lhoestq** [comment]
Created an issue here: https://github.com/huggingface/datasets/issues/4152

File: datasets/biwi_kinect_head_pose/README.md [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+- a corresponding rgb image (both 640x480 pixels),
+- annotation
+
+The head pose range covers about +-75 degrees yaw and +-60 degrees pitch. The ground truth is the 3D location of the head and its rotation.
+
+### Data Processing
+
+Example code for reading a compressed binary depth image file provided by the authors.
+
+```cpp
```
[the remark this answers is not in the document]
Comment (dnaveenr) [reply datasets/biwi_kinect_head_pose/README.md:72]: The code isn't available online, its actually part of the downloaded dataset. We can collapse the code. I'll add this part.

**Comment by dnaveenr** [comment]
Got it. Thanks for the comments. I've collapsed the C++ code in the readme and added the suggestions.

**Comment by dnaveenr** [comment]
Got it. Thanks for the comments. I've collapsed the C++ code in the readme and added the suggestions.

**Comment by lhoestq** [comment]
Hi ! The `AttributeError ` bug has been fixed, feel free to merge `master` into your branch ;)

**Comment by dnaveenr** [comment]
I haven't been able to figure out why CI is failing, the error shown is : 

> E           ValueError: The following issues have been found in the dataset cards:
> E           README Parsing:
> E           list index out of range
> E           The following issues have been found in the dataset cards:
> E           README Validation:
> E           list index out of range

Any inputs would be helpful.

**Comment by lhoestq** [comment]
I think it's because there are tabulations in the c code, can you replace them with regular spaces please ?

File: datasets/biwi_kinect_head_pose/README.md [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+  author = {Fanelli, Gabriele and Dantone, Matthias and Gall, Juergen and Fossati, Andrea and Van Gool, Luc},
+  title = {Random Forests for Real Time 3D Face Analysis},
+  journal = {Int. J. Comput. Vision},
+  year = {2013},
+  month = {February},
+  volume = {101},
+  number = {3},
+  pages = {437--458}
+}
+```
```
Comment (lhoestq) [inline datasets/biwi_kinect_head_pose/README.md:304]:
````suggestion
```

### Contributions

Thanks to [@dnaveenr](https://github.com/dnaveenr) for adding this dataset.
````

**Comment by dnaveenr** [comment]
@lhoestq , initially the idea was to have one example = one image with an additional field mentioning the frame_number. But each subject, we had a head template, calibration information for the depth and the color camera which was common to all the examples for that subject. Also, the images were continuous frames.
@mariosasko  suggested this structure and it made sense to group the images together for a particular subject.

**Comment by dnaveenr** [comment]
> Don't you think it would be more practical to have one example = one image in this dataset ?

Having one example = one image would be good but since we have a head template, calibration information for the depth and the color camera which is common to all the images for that subject and the images being continuous frames, I think it makes sense to group the images together for each subject. This will make the feature representation easier.

**Comment by lhoestq** [comment]
Ok I see, sounds good then. Users can still separate the images if they want to

**Review by lhoestq — APPROVED** [review]
It's all good then, thanks for adding this dataset ! :)

**Comment by lhoestq** [comment]
The CI fails are unrelated to this PR and fixed on master, merging !

**Comment by dnaveenr** [comment]
Great. Thanks @lhoestq , I think we can close this issue now. ( #3822 )

**Pull request merged**